### ISSUE TITLE
Please document how to run with LIBUSB_DEBUG=x or enable libusb debug output with nut debug options

### ISSUE THREAD

**Issue by tormodvolden** [networkupstools/nut#2616]
To investigate USB issues it is important to see the debug output from libusb. Either the env var LIBUSB_DEBUG must be set to e.g. 4, or the program can call libusb_set_debug() to set debug level per libusb context.

**Comment by jimklimov** [comment]
As a quick shot with existing builds - does it suffice to export the envvar in USB-capable driver units/init-scripts (commonly via `nut.conf`)?

For later versions, I'd envision something like what's done in recent years for `nut_debug_level` via `DEBUG_MIN` settings and real-time protocol settings to dial the knob in-vivo.

"Per context" here maps to one process's usage of libusb (from init to close), right?

**Comment by tormodvolden** [comment]
Yes, if there is a mechanism to set environment variables when executing the driver binaries, that will work.

Setting the env var on the command line when launching a program is technically easy but there are syntax differences between operating systems (or rather their shells), and when things e.g. run via systemd it becomes more difficult for many end-users.

Per-context refers to the libusb concept of context. You can have multiple libusb contexts in one process without them interfering with each other. They will for instance have their own list of discovered devices and their own set of callbacks for events, their own debug level and so forth. This makes sense e.g. for a larger program with plugins or modules that do USB but operate independently.

**Comment by jimklimov** [comment]
Got it, thanks!

**Comment by jimklimov** [comment]
For starters, linking docs to https://libusb.sourceforge.io/api-1.0/group__libusb__lib.html#ga2d6144203f0fc6d373677f6e2e89d2d2

**Comment by masterwishx** [comment]
`exported LIBUSB_DEBUG = 4` is not working yet ?

**Comment by jimklimov** [comment]
For environment variable setup through `nut.conf` file sourced for service startup (including drivers as a service - but not used directly as a program nor via `upsdrvctl`), either way - for a shell interpreter you would

````
export LIBUSB_DEBUG=4
````

Not `exported`, no spaces around `=` ...

**Comment by masterwishx** [comment]
> For environment variable setup through `nut.conf` file sourced for service startup (including drivers as a service - but not used directly as a program nor via `upsdrvctl`), either way - for a shell interpreter you would
> 
> ```
> export LIBUSB_DEBUG=4
> ```
> 
> Not `exported`, no spaces around `=` ...

In nut plugin setup config all settings with spaces, but not a problem I can check both with and without spaces.
So libusb debug work with debug_min=6 also?

**Comment by jimklimov** [comment]
We here are making NUT, not its HA plugin, sorry ;)

From what I saw, it is set up by yaml files that nut configs get generated from, not by nut configs directly. Maybe you should "cut out the middleman" and run NUT programs directly while investigating?..

AFAIK the plugin uses packaged (or custom-built, per wiki) NUT binaries and scripts etc., so maybe you can inject the `EnvironmentVariable=LIBUSB_DEBUG=4` to the `nut-driver@.service` definition temporarily - assuming plugin rewrites `nut.conf` so you can't really edit it (and `chattr` to make it immutable is not helpful either)?..

And "no", the `debug_min` setting for NUT verbosity is not currently linked to libusb's own debug. Perhaps adding it as a setting to be recognized from `ups.conf` and causing a `setenv()` early in `drivers/main.c` (rather than diligent setting of libusb context verbosity, as proposed in that issue originally) is also a decent quick solution.

**Comment by masterwishx** [comment]
> We here are making NUT, not its HA plugin, sorry ;)
> 
> From what I saw, it is set up by yaml files that NUT `*.conf` files get generated from, not by editing NUT configs in UI directly. Maybe you should "cut out the middleman" and run NUT programs directly while investigating?..
> 
> AFAIK the plugin uses packaged (or custom-built, per wiki) NUT binaries and scripts etc., so maybe you can inject the `EnvironmentVariable=LIBUSB_DEBUG=4` to the `nut-driver@.service` definition temporarily - assuming plugin rewrites `nut.conf` so you can't really edit it (and `chattr` to make it immutable is not helpful either)?..
> 
> And "no", the `debug_min` setting for NUT verbosity is not currently linked to libusb's own debug. Perhaps adding it as a setting to be recognized from `ups.conf` and causing a `setenv()` early in `drivers/main.c` (rather than diligent setting of libusb context verbosity, as proposed in that issue originally) is also a decent quick solution.

Sure, You are 100% right. 
BTW, I'm on Unraid plugin for NUT not HA :).

**Comment by jimklimov** [comment]
Oops :)

**Comment by jimklimov** [comment]
At the moment, this is supported in code and documented among NUT USB vars, I guess the ticket's goal is achieved.

```
$ git grep LIBUSB_DEBUG
NEWS.adoc:   * ...should now accept a `LIBUSB_DEBUG=INTEGER` setting in `ups.conf`
conf/ups.conf.sample:# LIBUSB_DEBUG: OPTIONAL.  For run-time troubleshooting of USB-capable NUT
docs/man/nut_usb_addvars.txt:*LIBUSB_DEBUG =* 'INTEGER'::
docs/man/nut_usb_addvars.txt:For the latter, you can set the `LIBUSB_DEBUG` driver option; alternatively
docs/man/nut_usb_addvars.txt:you can classically export the environment variable `LIBUSB_DEBUG` before
docs/man/ups.conf.txt:*LIBUSB_DEBUG* 'INTEGER'::
docs/man/ups.conf.txt:*LIBUSB_DEBUG* 'INTEGER'::
drivers/main.c: if (!strcmp(var, "LIBUSB_DEBUG")) {
drivers/main.c:         char buf[SMALLBUF], *s = getenv("LIBUSB_DEBUG");
drivers/main.c:                 upslogx(LOG_INFO, "WARNING : Invalid LIBUSB_DEBUG value found in ups.conf for the driver, defaulting to %d", lvl);
drivers/main.c:                 upsdebugx(msglvl, "Old value of LIBUSB_DEBUG=%s in envvars", s);
drivers/main.c:         upsdebugx(msglvl, "Enabling LIBUSB_DEBUG=%s from ups.conf", buf);
drivers/main.c:         setenv("LIBUSB_DEBUG", buf, 1);
drivers/main.c: if (!strcmp(var, "LIBUSB_DEBUG")) {
drivers/main.c:         char *s = getenv("LIBUSB_DEBUG");
drivers/main.c:                 upslogx(LOG_INFO, "WARNING : Invalid LIBUSB_DEBUG value found in ups.conf for the driver, defaulting to %d", lvl);
drivers/main.c:                 upsdebugx(msglvl, "Old value of LIBUSB_DEBUG=%s in envvars", s);
drivers/main.c:         upsdebugx(msglvl, "Enabling LIBUSB_DEBUG=%s from ups.conf", buf);
drivers/main.c:         setenv("LIBUSB_DEBUG", buf, 1);
include/nutconf.hpp:    inline long long int getLibusbDebug()   const { return getInt("LIBUSB_DEBUG"); }
include/nutconf.hpp:    inline void setLibusbDebug(long long int num)       { setInt("LIBUSB_DEBUG",  num); }
include/nutconf.hpp:    inline long long int getLibusbDebug(const std::string & ups)               const { return getInt(ups, "LIBUSB_DEBUG"); }
include/nutconf.hpp:    inline void setLibusbDebug(const std::string & ups, long long int val)                    { setInt(ups, "LIBUSB_DEBUG",        val); }
scripts/augeas/nutupsconf.aug.tpl:                 | "LIBUSB_DEBUG"
```
